File: ll.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None

    def __str__(self):
        return str(self.value)

    def set_next(self, value):
        self.next = Node(value)
        return self.next


class LinkedList:
    def __init__(self, head=None):
        self.head = head

    def __str__(self):
        values = []
        current = self.head
        while current:
            values.append(str(current))
            current = current.next

        return " -> ".join(values)

    def reverse(self):
        previous = None
        current = self.head

        while current.next:
            tmp = current.next
            current.next = previous
            previous = current
            current = tmp

        current.next = previous

        self.head = current


def main():
    head = Node("a")
    head.set_next("b").set_next("c").set_next("d").set_next("e")

    ll = LinkedList(head)
    print(ll)
    ll.reverse()
    print(ll)


if __name__ == "__main__":
    main()

File: ll.js
class Node {
    constructor(val) {
        this.val = val;
        this.next = null;
    }
}

const a = new Node('A');
const b = new Node('B');
const c = new Node('C');
const d = new Node('D');


a.next = b;
b.next = c; 
c.next = d;
// d.next = null ... A -> B -> C -> D -> null

const printLinkedList = (head) => {
    let current = head;
    while (current != null) {
        console.log(current.val);
        current = current.next;
    }
};

printLinkedList(a);


